src/main/php/inject/Injector.class.php

```diff
@@ -82,59 +82,28 @@ public function bind($type, $impl, $name= null) {
     } else {
       $this->bindings[$t->literal()][$name]= $impl instanceof Binding ? $impl : self::asBinding($t, $impl);
     }
+
     return $this;
   }
 
   /**
-   * Get a binding
+   * Returns the lookup if it provides a value, null otherwise
    *
-   * @param  string|lang.Type $type
-   * @param  string $name
-   * @return var or NULL if none exists
+   * @param  inject.Binding $lookup
+   * @return ?inject.Binding
    */
-  public function get($type, $name= null) {
-    $t= $type instanceof Type ? $type : Type::forName($type);
-
-    // Prevent lookup loops, see https://github.com/xp-forge/inject/issues/24
-    $key= $t->getName().'@'.$name;
-    if (isset($this->protect[$key])) throw new IllegalArgumentException('Lookup loop created by '.$key);
-
-    try {
-      $this->protect[$key]= true;
-      if ($t instanceof TypeUnion) {
-        foreach ($t->types() as $type) {
-          if ($instance= $this->get($type, $name)) return $instance;
-        }
-      } else if ($t instanceof Nullable) {
-        return $this->get($t->underlyingType(), $name);
-      } else if (self::$PROVIDER->isAssignableFrom($t)) {
-        $literal= $t->genericArguments()[0]->literal();
-        if (isset($this->bindings[$literal][$name])) {
-          return $this->bindings[$literal][$name]->provider($this);
-        }
-      } else {
-        $literal= $t->literal();
-        if (isset($this->bindings[$literal][$name])) {
-          return $this->bindings[$literal][$name]->resolve($this);
-        } else if (null === $name && $t instanceof XPClass && !($t->isInterface() || $t->getModifiers() & MODIFIER_ABSTRACT)) {
-          return $this->newInstance($t);
-        }
-      }
-      return null;
-    } finally {
-      unset($this->protect[$key]);
-    }
+  private function provided($lookup) {
```

Comment:
This should be inside `inject.Binding` but would mean a BC break. Too bad we don't have [default methods](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods) in interfaces